### PR TITLE
Fix incorrect assumption around the presence of ICF frames in EH codebase for 64-bit targets

### PR DESCRIPTION
**Pull request by fadimounir**
Fix for https://github.com/dotnet/runtime/issues/34524.

The main issue here is a disconnect between what the JIT produces, and the assumption we had around it in the EH code:
https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/exceptionhandling.cpp#L1824

Clearly this is not what the JIT does:
https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lower.cpp#L3497

This means that it is not correct to limit this define to only 32-bit targets:
https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/exceptionhandling.h#L16

To fix this issue, we should have one standard for all targets (which is the case today anyways): The ICF is always initialized and linked in the prolog of any method that contains PInvokes, and unlinked at the method epilog for jitted code (R2R is a slightly different story)

### CODE REVIEW

**Comment by jkotas** [comment]
> The ICF is always initialized and linked in the prolog of any method that contains PInvokes, and unlinked at the method epilog for jitted code (R2R is a slightly different story)

I thought that we agree on that link / unlink around PInvoke is the right thing to standardize on (ie same as R2R).

We can initialize the frame in the prolog or right before the PInvoke, but that is just a local detail - it does not affect how exception handling works.

File: src/coreclr/src/vm/exceptionhandling.h
Comment (jkotas) [inline src/coreclr/src/vm/exceptionhandling.h:17]: This will need matching changes in the JIT as well. You should see a functional test failures without matching changes in the JIT.

**Comment by janvorli** [comment]
> I thought that we agree on that link / unlink around PInvoke is the right thing to standardize on (ie same as R2R).

I would also prefer that. The presence of linked but inactive InlinedCallFrame is just confusing.

File: src/coreclr/src/vm/exceptionhandling.h
Comment (fadimounir) [reply src/coreclr/src/vm/exceptionhandling.h:17]: The JIT today links/unlinks the ICF in the method prolog/epilog, and activates/deactivates the ICF around the pinvoke call. It does this for both 32-bits and 64-bits. The fix here is to match the behavior in the EH codebase.

As a second step after this fix is in, I can look into merging the linking+activation in one step around the pinvoke call (this will match the behavior of R2R code as well), but that would be a slightly bigger change I believe

**Comment by fadimounir** [comment]
As a second step after this fix is in, I can look into merging the linking+activation in one step around the pinvoke call (this will match the behavior of R2R code as well), but that would be a slightly bigger change I believe. At this point, I want to fix the bug first (to also unblock my other PR that is failing with gcstress runs because of this)

**Comment by fadimounir** [comment]
gcstress results are comparable to a baseline I ran a few days ago: https://dev.azure.com/dnceng/public/_build/results?buildId=582616&view=ms.vss-test-web.build-test-results-tab

We can merge these changes at this point to unblock https://github.com/dotnet/runtime/pull/33733, then I'll make the changes to activate+link the ICF around the pinvoke calls.

@jkotas @janvorli Does this plan sound good?

**Review by jkotas — APPROVED** [review]
Sounds good

**Pull request merged**
